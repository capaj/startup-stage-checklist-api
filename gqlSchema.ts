import {
  SchemaRoot,
  Query,
  Mutation,
  Field,
  ObjectType,
  compileSchema,
  DuplexObjectType,
  DuplexField,
  QueryAndMutation,
} from 'decapi'
import { GraphQLInt } from 'graphql'

export class ConstructorAssigner<T = any> {
  constructor(parameters: Partial<T>) {
    Object.assign(this, parameters)
  }
}

@DuplexObjectType()
class Stage extends ConstructorAssigner {
  @Field({ type: GraphQLInt })
  id: number
  @DuplexField()
  title: string

  @DuplexField()
  completedAt: Date | null
}

@ObjectType()
class StageMutation extends Stage {
  @Field()
  edit(title: string): Stage {
    this.title = title
    return this
  }
  @Field()
  complete(): Stage {
    this.completedAt = new Date()
    return this
  }
  @Field()
  remove(): number {
    startupDB = startupDB.filter((s) => s.id !== this.id)
    return startupDB.length
  }
}

@DuplexObjectType()
class Startup extends ConstructorAssigner {
  @Field({ type: GraphQLInt })
  id: number
  @DuplexField()
  name: string

  @Field()
  stages: StageMutation[]
}

let startupDB: Startup[] = [
  {
    id: 1,
    name: 'Crave tech',
    stages: [
      new StageMutation({ id: 1, name: 'Lord of the Rings' }),
      new StageMutation({ id: 2, name: 'Harry Potter' }),
    ],
  },
]

@SchemaRoot()
class StartupStageChecklistSchema {
  @QueryAndMutation()
  startup(id: number): Startup | undefined {
    return startupDB.find(({ id: sId }) => sId === id)
  }

  @Query({ type: [Startup] })
  startups(): Startup[] {
    return startupDB
  }
}

export const gqlSchema = compileSchema(StartupStageChecklistSchema)
