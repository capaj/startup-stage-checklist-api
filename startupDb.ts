import { TodoMutation, Todo, StartupMutation } from './gqlSchema'
import { genId } from './genId'
export const startupDB: StartupMutation[] = [
  new StartupMutation({
    id: genId(),
    name: 'Crave tech',
    stages: [
      new TodoMutation({
        id: genId(),
        title: 'Foundation',
        todos: [
          new Todo({ id: genId(), title: 'Setup virtual office' }),
          new Todo({ id: genId(), title: 'Set mission and vision' }),
          new Todo({ id: genId(), title: 'select business name' }),
          new Todo({ id: genId(), title: 'buy domains' }),
        ],
      }),
      new TodoMutation({
        id: genId(),
        title: 'Discovery',
        todos: [
          new Todo({ id: genId(), title: 'Roadmap' }),
          new Todo({ id: genId(), title: 'Competitor analysis' }),
        ],
      }),
      new TodoMutation({
        id: genId(),
        title: 'Delivery',
        todos: [
          new Todo({ id: genId(), title: 'Mvp' }),
          new Todo({ id: genId(), title: 'Launch' }),
        ],
      }),
    ],
  }),
]

console.log(startupDB[0].id)
