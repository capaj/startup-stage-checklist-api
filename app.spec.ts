import { gqlSchema } from './gqlSchema'
import { graphql } from 'graphql'
import { startupDB } from './startupDb'

describe('suite name', () => {
  it('should query all data about startup, stages and todos', async () => {
    const res = await graphql({
      contextValue: { db: startupDB },
      schema: gqlSchema,
      source: `{
        startups {
          id
          name
          stages {
            id
            title
            completed
            todos {
              id
              title
              completedAt
            }
          }
        }
      }
    `,
    })
    expect(res).toMatchSnapshot()
  })

  it
})
