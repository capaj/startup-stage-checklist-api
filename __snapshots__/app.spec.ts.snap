// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`suite name should query all data about startup, stages and todos 1`] = `
Object {
  "data": Object {
    "startups": Array [
      Object {
        "id": "3e9",
        "name": "Crave tech",
        "stages": Array [
          Object {
            "completed": false,
            "id": "3ea",
            "title": "Foundation",
            "todos": Array [
              Object {
                "completedAt": null,
                "id": "3eb",
                "title": "Setup virtual office",
              },
              Object {
                "completedAt": null,
                "id": "3ec",
                "title": "Set mission and vision",
              },
              Object {
                "completedAt": null,
                "id": "3ed",
                "title": "select business name",
              },
              Object {
                "completedAt": null,
                "id": "3ee",
                "title": "buy domains",
              },
            ],
          },
          Object {
            "completed": false,
            "id": "3ef",
            "title": "Discovery",
            "todos": Array [
              Object {
                "completedAt": null,
                "id": "3f0",
                "title": "Roadmap",
              },
              Object {
                "completedAt": null,
                "id": "3f1",
                "title": "Competitor analysis",
              },
            ],
          },
          Object {
            "completed": false,
            "id": "3f2",
            "title": "Delivery",
            "todos": Array [
              Object {
                "completedAt": null,
                "id": "3f3",
                "title": "Mvp",
              },
              Object {
                "completedAt": null,
                "id": "3f4",
                "title": "Launch",
              },
            ],
          },
        ],
      },
    ],
  },
}
`;
